### ISSUE TITLE
Regression: generic inference fails for coroutine list passed as `list[Awaitable[T]]`

### ISSUE THREAD

**Issue by jelle-openai** [astral-sh/ty#4292]
### Summary

```python
from collections.abc import Awaitable
from typing import TypeVar

T = TypeVar("T")

async def gather(values: list[Awaitable[T]]) -> list[T]:
    return [await value for value in values]

async def produce() -> int:
    return 1

async def main() -> None:
    result = await gather([produce()])
    reveal_type(result)
```

ty 0.0.69, 0.0.70, and 0.0.71 accept this and infer `list[int]`. ty 0.0.72 instead reports:

```text
error[invalid-argument-type]: Expected list[Awaitable[Unknown]],
found list[CoroutineType[Any, Any, int]]
```

It also infers `result` as `list[Unknown]`. Other type checkers in multiplay infer `list[int]` too.

### Version

0.0.72

**Comment by carljm** [comment]
Thanks! Here's a non-async reproducer of the same issue:

```py
from typing import Protocol, reveal_type

class Source[T](Protocol):
    def get(self) -> T: ...

class IntSource:
    def get(self) -> int:
        return 1

def collect[T](values: list[Source[T]]) -> list[T]:
    return [value.get() for value in values]

reveal_type(collect([IntSource()]))
```

We now infer `[IntSource()]` as `list[IntSource]` and reject the assignment to `list[Source[T]]` due to list invariance.

The proximate cause is https://github.com/astral-sh/ruff/pull/27707, which starts handling variance correctly in generic inference. Before that, we wrongly always treated the `IntSource <: Source[T]` check as covariant, even within `list[...]`. In this case that happened to do the right thing, since a list literal can't be aliased and therefore can safely be treated covariantly during inference of its type. In general we handle this via type context / bidirectional inference. But our bidirectional inference currently discards type contexts that contain an unsolved type variable, so we do not use `list[Source[T]]` as bidirectional context when inferring a type for `[IntSource()]`.

I think the right solution here is https://github.com/astral-sh/ty/issues/3933, but I'll look into whether we can do something in the shorter term to avoid the regression.
